### PR TITLE
fix: `@aws-sdk/client-s3` instrumentation could cause `getSignedUrl` to crash

### PR DESCRIPTION
**Pull request by trentm**
Closes: https://github.com/elastic/apm-agent-nodejs/issues/3464

### CODE REVIEW

File: lib/instrumentation/modules/@aws-sdk/client-s3.js [hunk 1]
```diff
@@ -61,14 +61,15 @@ function s3MiddlewareFactory (client, agent) {
...
     {
       middleware: (next, context) => async (args) => {
-        const input = args.input
-        const bucket = input && input.Bucket
-        const resource = resourceFromBucket(bucket)
+        // Ensure there is a span from the wrapped `client.send()`.
         const span = agent._instrumentation.currSpan()
-
-        if (!span) {
+        if (!span || !(span.type === TYPE && span.subtype === SUBTYPE)) {
```
Comment (trentm) [inline lib/instrumentation/modules/@aws-sdk/client-s3.js:66]: Reviewer note: This is the fix for the first surprise. We do a slightly weak check that the current span is one likely created by our `client.send()` wrapper (in smithy-client.js).

File: lib/instrumentation/modules/@aws-sdk/client-s3.js [hunk 2]
```diff
@@ -152,17 +153,18 @@ function s3MiddlewareFactory (client, agent) {
...
-            port: context[elasticAPMStash].port,
             service: {
               name: SUBTYPE,
               type: TYPE
             }
           }
+          if (context[elasticAPMStash]) {
+            destContext.address = context[elasticAPMStash].hostname
+            destContext.port = context[elasticAPMStash].port
+          }
```
Comment (trentm) [inline lib/instrumentation/modules/@aws-sdk/client-s3.js:164]: Reviewer note: This is the fix for the second surprise. It is possible for middleware (if it doesn't call `next()`) to result in our "finalizeRequest" middleware not being called, so `context[elasticAPMStack]` wouldn't be set.

File: lib/instrumentation/modules/@aws-sdk/client-s3.js [hunk 3]
```diff
@@ -190,7 +192,6 @@ function s3MiddlewareFactory (client, agent) {
         }
 
         context[elasticAPMStash] = {
-          protocol: req.protocol,
```
Comment (trentm) [inline lib/instrumentation/modules/@aws-sdk/client-s3.js:193]: Reviewer note: This field wasn't being used, so I dropped it.

File: test/instrumentation/modules/@aws-sdk/fixtures/use-client-s3.js [hunk 4]
```diff
@@ -129,14 +152,30 @@ async function useClientS3 (s3Client, bucketName) {
   // https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/classes/getobjectcommand.html
   command = new GetObjectCommand({ Bucket: bucketName, Key: key })
   data = await s3Client.send(command)
-  log.info({ data }, 'getObject')
+  // `data.Body` is a *stream*, so we cannot just log it
+  // https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/getobjectcommandoutput.html#body
+  const body = await slurpStream(data.Body)
+  assert(body === content)
+  delete data.Body
+  log.info({ data, body }, 'getObject')
```
Comment (trentm) [inline test/instrumentation/modules/@aws-sdk/fixtures/use-client-s3.js:160]: Reviewer note: This change is unrelated to the bug. With aws-sdk **v3** that `data.Body` is a Node.js *stream* object. Logging it results in an unhelpful and huge log record. This changes to fully read that stream so we can log the S3 object content instead.

File: test/instrumentation/modules/@aws-sdk/fixtures/use-client-s3.js [hunk 5]
```diff
@@ -129,14 +152,30 @@ async function useClientS3 (s3Client, bucketName) {
...
+  // This is interesting to test, because `getSignedUrl` uses the command
+  // `middlewareStack` -- including our added middleware -- **without** calling
+  // `s3Client.send()`. The test here is to ensure this doesn't break.
+  const customSpan = apm.startSpan('get-signed-url')
+  const signedUrl = await getSignedUrl(
+    s3Client,
+    new GetObjectCommand({ Bucket: bucketName, Key: key }),
+    { expiresIn: 3600 })
+  log.info({ signedUrl }, 'getSignedUrl')
+  customSpan.end()
```
Comment (trentm) [inline test/instrumentation/modules/@aws-sdk/fixtures/use-client-s3.js:172]: Reviewer note: Adding a test case. Before the fixes above, this would throw.

File: test/instrumentation/modules/@aws-sdk/fixtures/use-client-s3.js [hunk 6]
```diff
@@ -228,7 +267,8 @@ function main () {
       s3Client.destroy()
       process.exitCode = 0
     },
-    function () {
+    function (err) {
+      apm.logger.error(err, 'useClientS3 rejected')
```
Comment (trentm) [inline test/instrumentation/modules/@aws-sdk/fixtures/use-client-s3.js:271]: Reviewer note: Before this change the error, if any, was silently not logged -- making debugging more difficult.

**Review by david-luna — APPROVED** [review]
